### PR TITLE
udev: use link info obtained through netlink

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Comment by yuwata** [comment]
Now this is ready for review. PTAL.

File: src/libsystemd/sd-netlink/netlink-message.c [hunk 1]
```diff
@@ -1093,38 +1093,39 @@ int sd_netlink_message_read_strv(sd_netlink_message *m, unsigned short container
...
 }
 
-static int netlink_container_parse(sd_netlink_message *m,
-                                   struct netlink_container *container,
-                                   struct rtattr *rta,
-                                   size_t rt_len) {
+static int netlink_container_parse(
+                sd_netlink_message *m,
+                struct netlink_container *container,
+                struct rtattr *rta,
```
Comment (poettering) [inline src/libsystemd/sd-netlink/netlink-message.c:1099]: shouldn't this be const?

File: src/libsystemd/sd-netlink/netlink-message.c [hunk 2]
```diff
@@ -1093,38 +1093,39 @@ int sd_netlink_message_read_strv(sd_netlink_message *m, unsigned short container
...
         unsigned short len = rt_len;
         for (; RTA_OK(rta, len); rta = RTA_NEXT(rta, len)) {
-                unsigned short type;
+                uint16_t attr;
 
-                type = RTA_TYPE(rta);
+                attr = RTA_TYPE(rta);
 
-                if (!GREEDY_REALLOC0(attributes, type + 1))
+                if (!GREEDY_REALLOC0(attributes, attr + 1))
```
Comment (poettering) [inline src/libsystemd/sd-netlink/netlink-message.c]: shouldn't this be:

```c
if (!GREEDY_REALLOC0(attributes, MAX(attr, max_attr) + 1))
```

i mean, yes, greedy realloc is not gonna shrink the memory, but it still looks cleaner to always specify the full amount memory we want to use here.

Comment (poettering) [reply src/libsystemd/sd-netlink/netlink-message.c]: hmm, what about overflows here? whta happens if attr is 0xffff? then the +1 here will overflow. We should probably catch that.

File: src/udev/udev-builtin-net_id.c [hunk 3]
```diff
@@ -97,19 +92,19 @@ static sd_device *skip_virtio(sd_device *dev) {
         return dev;
 }
 
-static int get_virtfn_info(sd_device *dev, struct netnames *names, struct virtfn_info *ret) {
+static int get_virtfn_info(sd_device *pcidev, sd_device **ret_physfn_pcidev, char **ret_suffix) {
         _cleanup_(sd_device_unrefp) sd_device *physfn_pcidev = NULL;
         const char *physfn_syspath, *syspath;
-        struct dirent *dent;
+        _cleanup_free_ char *suffix = NULL;
```
Comment (poettering) [inline src/udev/udev-builtin-net_id.c]: move to innermost scope?

File: src/udev/udev-builtin-net_id.c [hunk 4]
```diff
@@ -142,19 +139,16 @@ static int get_virtfn_info(sd_device *dev, struct netnames *names, struct virtfn
                         continue;
 
                 if (streq(syspath, virtfn_pci_syspath)) {
-                        if (!snprintf_ok(suffix, sizeof(suffix), "v%s", &dent->d_name[6]))
-                                return -ENOENT;
+                        if (asprintf(&suffix, "v%s", n) < 0)
+                                return -ENOMEM;
```
Comment (poettering) [inline src/udev/udev-builtin-net_id.c:144]: for concatenating strings strjoin() is like a million times more efficient. (and morea readable, too...). 

hence:

```c
suffix = strjoin("v", n);
…
```

File: src/udev/udev-builtin-net_id.c [hunk 5]
```diff
@@ -39,7 +39,7 @@
 #define ONBOARD_14BIT_INDEX_MAX ((1U << 14) - 1)
 #define ONBOARD_16BIT_INDEX_MAX ((1U << 16) - 1)
 
-enum netname_type{
+typedef enum NetNameType{
```
Comment (poettering) [inline src/udev/udev-builtin-net_id.c]: space before { ?

**Comment by poettering** [comment]
looks good, just some minor comments, see above

File: src/libsystemd/sd-netlink/netlink-message.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd/sd-netlink/netlink-message.c:1099]: Unfortunately, `RTA_NEXT()` macro drops `const` qualifier.

**Comment by yuwata** [comment]
@poettering Thank you for your review. All comments, except for 'const' qualifier, are addressed. PTAL.

**Pull request merged**